core/graphelier-service/models/orderbooks.go

```diff
@@ -0,0 +1,188 @@
+type Orderbook struct {
+	Instrument string   `json:"instrument"`
+	Bids       []*Level `json:"bids"`
+	Asks       []*Level `json:"asks"`
+	Timestamp  uint64   `json:"timestamp"`
+}
+
+// ApplyMessagesToOrderbook : Applies each individual message to the orderbook
+func (orderbook *Orderbook) ApplyMessagesToOrderbook(messages []*Message) *Orderbook {
+	for _, message := range messages {
+		if message.OrderID == 0 {
+			continue
+		}
+		switch MessageHandlers(message.MessageType) {
+		case NewOrder:
+			orderbook.applyNewOrder(message)
+		case Modify:
+			orderbook.applyModify(message)
+		case Delete:
+			orderbook.applyDelete(message)
+		case Execute:
+			orderbook.applyExecute(message)
+		case Ignore:
+			// Pass
+		}
+		orderbook.Timestamp = message.Timestamp
+	}
+
+	return orderbook
+}
+
+func (orderbook *Orderbook) applyNewOrder(message *Message) {
+	order := &Order{message.OrderID, message.ShareQuantity}
+	if message.Direction == -1 {
+		index, found := getLevelIndex(orderbook.Asks, message)
+		if !found {
+			level := &Level{message.Price, []*Order{order}}
+			orderbook.Asks = append(orderbook.Asks, nil)
+			copy(orderbook.Asks[index+1:], orderbook.Asks[index:])
+			orderbook.Asks[index] = level
+		} else {
+			orderbook.Asks[index].Orders = append(orderbook.Asks[index].Orders, order)
+		}
+	} else if message.Direction == 1 {
+		index, found := getLevelIndex(orderbook.Bids, message)
+		if !found {
+			level := &Level{message.Price, []*Order{order}}
+			orderbook.Bids = append(orderbook.Bids, nil)
+			copy(orderbook.Bids[index+1:], orderbook.Bids[index:])
+			orderbook.Bids[index] = level
+		} else {
+			orderbook.Bids[index].Orders = append(orderbook.Bids[index].Orders, order)
+		}
+	}
+}
+
+func (orderbook *Orderbook) applyModify(message *Message) {
+	orders, orderIndex := getOrders(orderbook, message)
+	if orderIndex < 0 || orders == nil {
+		return
+	}
+
+	o := (*orders)[orderIndex]
+	if message.ShareQuantity < o.Quantity {
+		o.Quantity -= message.ShareQuantity
+	} else {
+		orderbook.applyDelete(message)
+		message.ShareQuantity = o.Quantity - message.ShareQuantity
+		orderbook.applyNewOrder(message)
+	}
+}
+
+func (orderbook *Orderbook) applyDelete(message *Message) {
+	orders, orderIndex := getOrders(orderbook, message)
+	if orderIndex < 0 || orders == nil {
+		return
+	}
+	copy((*orders)[orderIndex:], (*orders)[orderIndex+1:])
+	(*orders)[len(*orders)-1] = nil
+	*orders = (*orders)[:len(*orders)-1]
+	if len(*orders) == 0 {
+		orderbook.removeEmptyLevels(message)
+	}
+}
+
+func (orderbook *Orderbook) applyExecute(message *Message) {
+	orders, orderIndex := getOrders(orderbook, message)
+	if orderIndex < 0 || orders == nil {
+		return
+	}
+	if (*orders)[orderIndex].Quantity > message.ShareQuantity {
+		(*orders)[orderIndex].Quantity -= message.ShareQuantity
+	} else {
+		orderbook.applyDelete(message)
+	}
+}
+
+func (orderbook *Orderbook) removeEmptyLevels(message *Message) {
```

Comment:
The name here might be a bit confusing as we are only removing a single empty level